File: tests/node/test_content_store_streaming.mjs
/**
 * Streaming-download tests for the Node.js `ContentStore` binding.
 *
 * Exercises the `ContentStore.fetchStream` wrapper across the three
 * construction paths (custom callback, `ContentStore` subclass, and
 * the built-in `localFile` backend) without touching any external
 * service. Covers correctness, error propagation, and a basic
 * backpressure smoke check.
 *
 * The runtime expectation is:
 *
 *     const iter = await store.fetchStream(handle);
 *     for await (const chunk of iter) {
 *         ...
 *     }
 */

import test from "ava";
import { mkdtempSync, rmSync } from "node:fs";
import { tmpdir } from "node:os";
import { join } from "node:path";

import { ContentStore } from "../../crates/blazen-node/index.js";

// =========================================================================
// Helpers
// =========================================================================

/** A throwaway handle the custom-store callbacks below ignore. */
function placeholderHandle() {
  return { id: "test-handle", kind: "other" };
}

/** Drain an async iterable into an array of Buffers. */
async function drain(iter) {
  const chunks = [];
  for await (const chunk of iter) {
    chunks.push(Buffer.from(chunk));
  }
  return chunks;
}

// =========================================================================
// 1. Custom store with bytes-returning fetchStream (legacy shape).
// =========================================================================

test("ContentStore.fetchStream · custom store returning Buffer collapses to one chunk", async (t) => {
  const expected = Buffer.from("hello world");

  const store = ContentStore.custom({
    put: async (_body, _hint) => placeholderHandle(),
    resolve: async (_handle) => ({ type: "url", url: "https://example.invalid/blob" }),
    fetchBytes: async (_handle) => expected,
    fetchStream: async (_handle) => expected,
    name: "bytes-stream",
  });

  const handle = await store.put(Buffer.from("placeholder"), {});
  const iter = await store.fetchStream(handle);
  const chunks = await drain(iter);

  t.is(chunks.length, 1);
  t.is(chunks[0].toString("utf8"), "hello world");
});

// =========================================================================
// 2. Custom store with AsyncIterable-returning fetchStream (new shape).
// =========================================================================

test("ContentStore.fetchStream · custom store returning async generator yields chunk-by-chunk", async (t) => {
  const payloadChunks = [
    Buffer.from("hello "),
    Buffer.from("big "),
    Buffer.from("world"),
  ];

  const store = ContentStore.custom({
    put: async (_body, _hint) => placeholderHandle(),
    resolve: async (_handle) => ({ type: "url", url: "https://example.invalid/blob" }),
    fetchBytes: async (_handle) => Buffer.concat(payloadChunks),
    fetchStream: async (_handle) => {
      async function* gen() {
        for (const chunk of payloadChunks) {
          yield new Uint8Array(chunk);
        }
      }
      return gen();
    },
    name: "gen-stream",
  });

  const handle = await store.put(Buffer.from("placeholder"), {});
  const iter = await store.fetchStream(handle);
  const chunks = await drain(iter);

  t.is(chunks.length, 3);
  t.is(Buffer.concat(chunks).toString("utf8"), "hello big world");
  t.is(
    chunks.reduce((n, c) => n + c.length, 0),
    "hello big world".length,
  );
});

// =========================================================================
// 3. Subclass with async-generator fetchStream override.
// =========================================================================

test("ContentStore.fetchStream · subclass async-generator override yields each chunk", async (t) => {
  class MyStore extends ContentStore {
    constructor() {
      super();
    }

    async put(_body, _hint) {
      return placeholderHandle();
    }

    async resolve(_handle) {
      return { type: "url", url: "https://example.invalid/blob" };
    }

    async fetchBytes(_handle) {
      return Buffer.from("ab");
    }

    async *fetchStream(_handle) {
      yield new Uint8Array(Buffer.from("a"));
      yield new Uint8Array(Buffer.from("b"));
    }
  }

  const store = new MyStore();
  const handle = placeholderHandle();
  const iter = await store.fetchStream(handle);
  const chunks = await drain(iter);

  t.is(chunks.length, 2);
  t.is(chunks[0].toString("utf8"), "a");
  t.is(chunks[1].toString("utf8"), "b");
});

// =========================================================================
// 4. Built-in localFile round-trip.
// =========================================================================

test("ContentStore.fetchStream · localFile round-trip preserves the full payload", async (t) => {
  const root = mkdtempSync(join(tmpdir(), "blazen-fetch-stream-"));
  t.teardown(() => rmSync(root, { recursive: true, force: true }));

  const payloadSize = 100_000;
  const payload = Buffer.alloc(payloadSize, 1);

  const store = ContentStore.localFile(root);
  const handle = await store.put(payload, {});

  const iter = await store.fetchStream(handle);
  const chunks = await drain(iter);

  const total = chunks.reduce((n, c) => n + c.length, 0);
  t.is(total, payloadSize);
  t.true(Buffer.concat(chunks).equals(payload));
});

// =========================================================================
// 5. Error path — async generator throws mid-iteration.
// =========================================================================

test("ContentStore.fetchStream · errors raised mid-stream propagate to the consumer", async (t) => {
  class FailingStore extends ContentStore {
    constructor() {
      super();
    }

    async put(_body, _hint) {
      return placeholderHandle();
    }

    async resolve(_handle) {
      return { type: "url", url: "https://example.invalid/blob" };
    }

    async fetchBytes(_handle) {
      throw new Error("not used");
    }

    async *fetchStream(_handle) {
      yield new Uint8Array(Buffer.from("first chunk"));
      throw new Error("boom inside fetchStream");
    }
  }

  const store = new FailingStore();
  const handle = placeholderHandle();
  const iter = await store.fetchStream(handle);

  const collected = [];
  const error = await t.throwsAsync(async () => {
    for await (const chunk of iter) {
      collected.push(Buffer.from(chunk));
    }
  });

  // Don't pin the concrete error class — the bridge may wrap it as
  // ProviderError / generic Error depending on dispatch path. We do
  // assert the underlying message bubbled through and that the first
  // chunk arrived before the failure.
  t.truthy(error);
  t.regex(String(error.message ?? error), /boom inside fetchStream/);
  t.is(collected.length, 1);
  t.is(collected[0].toString("utf8"), "first chunk");
});

// =========================================================================
// 6. Backpressure smoke — many small chunks, slow consumer, ordered delivery.
// =========================================================================

test("ContentStore.fetchStream · backpressure smoke (32 × 16 KiB chunks, slow consumer)", async (t) => {
  const chunkCount = 32;
  const chunkSize = 16 * 1024;
  const expectedChunks = Array.from({ length: chunkCount }, (_, i) =>
    Buffer.alloc(chunkSize, i % 256),
  );

  const store = ContentStore.custom({
    put: async (_body, _hint) => placeholderHandle(),
    resolve: async (_handle) => ({ type: "url", url: "https://example.invalid/blob" }),
    fetchBytes: async (_handle) => Buffer.concat(expectedChunks),
    fetchStream: async (_handle) => {
      async function* gen() {
        for (const chunk of expectedChunks) {
          yield new Uint8Array(chunk);
          // Cooperative yield so producer and consumer interleave.
          await new Promise((resolve) => setImmediate(resolve));
        }
      }
      return gen();
    },
    name: "backpressure",
  });

  const handle = await store.put(Buffer.from("placeholder"), {});
  const iter = await store.fetchStream(handle);

  const received = [];
  for await (const chunk of iter) {
    received.push(Buffer.from(chunk));
    // Cooperative consumer pause — exercises the bounded pipe between
    // the JS-side producer and the Rust-side stream wrapper.
    await new Promise((resolve) => setImmediate(resolve));
  }

  t.is(received.length, chunkCount);
  t.is(
    received.reduce((n, c) => n + c.length, 0),
    chunkCount * chunkSize,
  );
  for (let i = 0; i < chunkCount; i += 1) {
    t.true(
      received[i].equals(expectedChunks[i]),
      `chunk ${i} mismatched (out-of-order or corrupted)`,
    );
  }
});

File: tests/python/test_content_store_streaming.py
"""Streaming-download tests for the Python ``ContentStore`` binding.

Exercises the ``ContentStore.fetch_stream`` wrapper across the three
construction paths (``custom`` callback, ``ContentStore`` subclass, and
the built-in ``local_file`` backend) without touching any external
service. Covers correctness, error propagation, and a basic
backpressure smoke check.

The runtime expectation is::

    iter = await store.fetch_stream(handle)
    async for chunk in iter:
        ...

(Equivalently: ``async for chunk in await store.fetch_stream(handle):``.)
"""

from __future__ import annotations

import asyncio
from collections.abc import AsyncIterator

import pytest

from blazen import ContentHandle, ContentKind, ContentStore

pytestmark = pytest.mark.asyncio


# ---------------------------------------------------------------------------
# Test helpers
# ---------------------------------------------------------------------------


def _placeholder_handle() -> ContentHandle:
    """Return a freshly constructed ``ContentHandle`` for callback tests.

    The custom-store callbacks below ignore the handle's identity entirely;
    they hand back a canned response. We only need *some* valid handle so
    the runtime has a stable wire identity to round-trip.
    """

    return ContentHandle("test-handle", ContentKind.Other)


async def _drain(iter_obj: object) -> list[bytes]:
    """Collect every chunk an async-iterator yields into a list."""

    chunks: list[bytes] = []
    async for chunk in iter_obj:  # type: ignore[attr-defined]
        chunks.append(chunk)
    return chunks


# ---------------------------------------------------------------------------
# 1. Custom store with bytes-returning ``fetch_stream`` (legacy shape).
# ---------------------------------------------------------------------------


async def test_custom_store_fetch_stream_returns_bytes() -> None:
    """A ``fetch_stream`` callback returning raw ``bytes`` is wrapped into
    a single-chunk async iterator on the Rust side."""

    expected = b"hello world"

    async def put(_body: dict, _hint: dict) -> ContentHandle:
        return _placeholder_handle()

    async def resolve(_handle: ContentHandle) -> dict:
        return {"type": "url", "url": "https://example.invalid/blob"}

    async def fetch_bytes(_handle: ContentHandle) -> bytes:
        return expected

    async def fetch_stream(_handle: ContentHandle) -> bytes:
        return expected

    store = ContentStore.custom(
        put=put,
        resolve=resolve,
        fetch_bytes=fetch_bytes,
        fetch_stream=fetch_stream,
        name="bytes-stream",
    )

    handle = await store.put(b"placeholder")
    iter_obj = await store.fetch_stream(handle)
    chunks = await _drain(iter_obj)

    assert chunks == [expected], chunks


# ---------------------------------------------------------------------------
# 2. Custom store with async-generator ``fetch_stream`` (new shape).
# ---------------------------------------------------------------------------


async def test_custom_store_fetch_stream_async_generator() -> None:
    """A ``fetch_stream`` callback returning an ``AsyncIterator[bytes]``
    streams chunk-by-chunk through the Rust bridge."""

    payload_chunks = [b"hello ", b"big ", b"world"]

    async def put(_body: dict, _hint: dict) -> ContentHandle:
        return _placeholder_handle()

    async def resolve(_handle: ContentHandle) -> dict:
        return {"type": "url", "url": "https://example.invalid/blob"}

    async def fetch_bytes(_handle: ContentHandle) -> bytes:
        return b"".join(payload_chunks)

    async def fetch_stream(_handle: ContentHandle) -> AsyncIterator[bytes]:
        async def gen() -> AsyncIterator[bytes]:
            for chunk in payload_chunks:
                yield chunk

        return gen()

    store = ContentStore.custom(
        put=put,
        resolve=resolve,
        fetch_bytes=fetch_bytes,
        fetch_stream=fetch_stream,
        name="gen-stream",
    )

    handle = await store.put(b"placeholder")
    iter_obj = await store.fetch_stream(handle)
    chunks = await _drain(iter_obj)

    assert len(chunks) == 3, chunks
    assert b"".join(chunks) == b"hello big world"


# ---------------------------------------------------------------------------
# 3. Subclass with async-generator ``fetch_stream`` override.
# ---------------------------------------------------------------------------


async def test_subclass_fetch_stream_async_generator() -> None:
    """A subclass that overrides ``fetch_stream`` as an async generator
    is driven through the same bridge as the callback path."""

    class MyStore(ContentStore):
        async def put(self, _body: dict, _hint: dict) -> ContentHandle:
            return _placeholder_handle()

        async def resolve(self, _handle: ContentHandle) -> dict:
            return {"type": "url", "url": "https://example.invalid/blob"}

        async def fetch_bytes(self, _handle: ContentHandle) -> bytes:
            return b"ab"

        async def fetch_stream(
            self, _handle: ContentHandle
        ) -> AsyncIterator[bytes]:
            yield b"a"
            yield b"b"

    store = MyStore()
    handle = _placeholder_handle()
    # MyStore.fetch_stream is an async generator function — calling it
    # returns an async-generator object directly (not a coroutine), so
    # we iterate without an outer ``await``.
    chunks: list[bytes] = []
    async for chunk in store.fetch_stream(handle):
        chunks.append(chunk)

    assert chunks == [b"a", b"b"], chunks


# ---------------------------------------------------------------------------
# 4. Built-in ``local_file`` round-trip.
# ---------------------------------------------------------------------------


async def test_local_file_fetch_stream_round_trip(tmp_path) -> None:
    """``local_file`` writes via ``put`` and reads back via
    ``fetch_stream``. We assert byte-for-byte equality without locking
    in any specific chunk count — ``ReaderStream`` chunking is an
    implementation detail."""

    payload = b"x" * 100_000

    store = ContentStore.local_file(tmp_path)
    handle = await store.put(payload)

    iter_obj = await store.fetch_stream(handle)
    chunks = await _drain(iter_obj)

    assert sum(len(c) for c in chunks) == len(payload)
    assert b"".join(chunks) == payload


# ---------------------------------------------------------------------------
# 5. Error path — generator raises mid-iteration.
# ---------------------------------------------------------------------------


async def test_subclass_fetch_stream_error_propagates() -> None:
    """An exception raised inside a subclass's ``fetch_stream`` async
    generator surfaces back to the Python consumer (via
    ``blazen_error_to_pyerr`` mapping)."""

    class FailingStore(ContentStore):
        async def put(self, _body: dict, _hint: dict) -> ContentHandle:
            return _placeholder_handle()

        async def resolve(self, _handle: ContentHandle) -> dict:
            return {"type": "url", "url": "https://example.invalid/blob"}

        async def fetch_bytes(self, _handle: ContentHandle) -> bytes:
            raise RuntimeError("not used")

        async def fetch_stream(
            self, _handle: ContentHandle
        ) -> AsyncIterator[bytes]:
            yield b"first chunk"
            raise RuntimeError("boom inside fetch_stream")

    store = FailingStore()
    handle = _placeholder_handle()
    # FailingStore.fetch_stream is an async generator — see the comment
    # in test_subclass_fetch_stream_async_generator above for why we do
    # not ``await`` it here.
    iter_obj = store.fetch_stream(handle)

    collected: list[bytes] = []
    with pytest.raises(Exception) as excinfo:
        async for chunk in iter_obj:
            collected.append(chunk)

    # We don't assert on the concrete exception type — the Blazen error
    # mapping may surface this as RuntimeError / ProviderError / similar
    # depending on the runtime build. We *do* assert the inner message
    # bubbled through and that the first chunk arrived before the error.
    assert "boom inside fetch_stream" in str(excinfo.value)
    assert collected == [b"first chunk"], collected


# ---------------------------------------------------------------------------
# 6. Backpressure smoke — many small chunks, slow consumer, ordered delivery.
# ---------------------------------------------------------------------------


async def test_fetch_stream_backpressure_smoke() -> None:
    """Producer yields 32 chunks of 16 KiB with cooperative yields between
    each; consumer awaits each ``__anext__`` with a tiny sleep. We verify
    every chunk arrives in order and the byte total matches.

    This is *correctness* under backpressure, not a memory benchmark —
    actual peak-RSS measurement is a future perf-suite concern."""

    chunk_count = 32
    chunk_size = 16 * 1024  # 16 KiB
    expected_chunks = [bytes([i % 256]) * chunk_size for i in range(chunk_count)]

    async def put(_body: dict, _hint: dict) -> ContentHandle:
        return _placeholder_handle()

    async def resolve(_handle: ContentHandle) -> dict:
        return {"type": "url", "url": "https://example.invalid/blob"}

    async def fetch_bytes(_handle: ContentHandle) -> bytes:
        return b"".join(expected_chunks)

    async def fetch_stream(_handle: ContentHandle) -> AsyncIterator[bytes]:
        async def gen() -> AsyncIterator[bytes]:
            for chunk in expected_chunks:
                yield chunk
                await asyncio.sleep(0)

        return gen()

    store = ContentStore.custom(
        put=put,
        resolve=resolve,
        fetch_bytes=fetch_bytes,
        fetch_stream=fetch_stream,
        name="backpressure",
    )

    handle = await store.put(b"placeholder")
    iter_obj = await store.fetch_stream(handle)

    received: list[bytes] = []
    aiter = iter_obj.__aiter__()
    while True:
        try:
            chunk = await aiter.__anext__()
        except StopAsyncIteration:
            break
        received.append(chunk)
        # Cooperative consumer pause — gives the producer task a chance to
        # fill the bounded mpsc channel and exercises the backpressure path.
        await asyncio.sleep(0)

    assert len(received) == chunk_count, len(received)
    assert received == expected_chunks
    assert sum(len(c) for c in received) == chunk_count * chunk_size
